mlflow/projects/__init__.py

```diff
@@ -93,41 +85,26 @@ def _resolve_experiment_id(experiment_name=None, experiment_id=None):
 
 
 def _run(uri, experiment_id, entry_point="main", version=None, parameters=None,
-         backend=None, backend_config=None, use_conda=True,
+         backend_name=None, backend_config=None, use_conda=True,
          storage_dir=None, synchronous=True, run_id=None):
     """
     Helper that delegates to the project-running method corresponding to the passed-in backend.
     Returns a ``SubmittedRun`` corresponding to the project run.
     """
-
-    parameters = parameters or {}
-    work_dir = _fetch_project(uri=uri, force_tempdir=False, version=version)
-    project = _project_spec.load_project(work_dir)
-    _validate_execution_environment(project, backend)
+    project, work_dir = fetch_and_validate_project(uri, version, entry_point, parameters)
+    _validate_execution_environment(project, backend_name)
     project.get_entry_point(entry_point)._validate_parameters(parameters)
-    if run_id:
-        active_run = tracking.MlflowClient().get_run(run_id)
-    else:
-        active_run = _create_run(uri, experiment_id, work_dir, entry_point)
-
-    # Consolidate parameters for logging.
-    # `storage_dir` is `None` since we want to log actual path not downloaded local path
-    entry_point_obj = project.get_entry_point(entry_point)
-    final_params, extra_params = entry_point_obj.compute_parameters(parameters, storage_dir=None)
-    for key, value in (list(final_params.items()) + list(extra_params.items())):
-        tracking.MlflowClient().log_param(active_run.info.run_id, key, value)
-
-    repo_url = _get_git_repo_url(work_dir)
-    if repo_url is not None:
-        for tag in [MLFLOW_GIT_REPO_URL, LEGACY_MLFLOW_GIT_REPO_URL]:
-            tracking.MlflowClient().set_tag(active_run.info.run_id, tag, repo_url)
-
-    # Add branch name tag if a branch is specified through -version
-    if _is_valid_branch_name(work_dir, version):
-        for tag in [MLFLOW_GIT_BRANCH, LEGACY_MLFLOW_GIT_BRANCH_NAME]:
-            tracking.MlflowClient().set_tag(active_run.info.run_id, tag, version)
-
-    if backend == "databricks":
+    active_run = get_or_create_run(run_id, uri, experiment_id, work_dir, entry_point)
```

Comment:
@jdlesage @smurching Had we come to a conclusion that each backend should be responsible for fetching project contents using predefined utility functions? (this is my preferred approach). If so, I think  we should move this logic into implementations of `backend.run()` rather than leaving them in `mlflow.projects._run`.

Reply:
Thanks @dbczumar & agreed, that was the conclusion we'd reached - would be good to move towards an implementation like that of the original PR ([link](https://github.com/mlflow/mlflow/pull/2415/files#diff-16104674981d54b31e320402a143eb19R100-R114)), where the backend has the option to fetch the project, log params, perform param validation, etc

Reply:
I created in utils some helper functions to fetch, validate, log and even tag. So backends can call them.
I did a mistake in my PR. I just pushed a change to fix it:  these functions are called in the client only if the backend is not managed by a plugin.
When all the existing backends will implement the abstract_class, I will remove all these calls in the `_run` method. As we agreed to split the PR (1 for abstract_class then 1 PR per backend), I will remove them in another PR.
Do you agree with this change ?